---
title: "Cluster quality control"
author: "Nadine Bestard"
date: "28/05/2021"
output:
  html_document:
    df_print: paged
editor_options: 
  markdown: 
    wrap: 80
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### set-up

```{r set-up, message=FALSE, warning=FALSE}
library(SingleCellExperiment)
library(here) # reproducible paths
library(scater) # plot reduced dims
library(dplyr) # df filtering
library(tibble)  # for `rownames_to_column`
library(scDblFinder) # detect doublets

```

```{r age}
age <- "old"
```

```{r load}
sce <- readRDS(here("processed", age, "sce_clusters_01.RDS"))
```

### Introduction

As in [publication] we will perform a cluster QC to remove clusters of poorer
quality. This will be assessed by the number of UMI counts, the mitochondrial
percentage, doublet analysis, ribosomal genes and the number of mice that
contribute to each cluster. Moreover we will keep in mind our experimental
groups in order to ensure biological effects are not being lost. To do so we use
a small cluster resolution, 5

```{r}
plotTSNE(sce, colour_by = "cluster_k5", point_size=0.5,
         point_alpha = 0.3, text_by = "cluster_k5", text_size = 3) +
  scale_color_hue() 
```

### Number of molecules per cluster

```{r}
plotTSNE(sce, colour_by = "total") + 
  ggtitle("Total number of umi counts per cell")
plotTSNE(sce, colour_by = "detected") +
  ggtitle("Detected number of genes per cell")
```

Lower values of umi counts and detected genes can be associated to lower quality
cells. Cells can also have lower expressed genes due to their biological state
or celltype.

```{r}
pct_cells<- 50
min_umi  <- 3000
```

Select clusters with `r pct_cells` % cells having less than `r min_umi`umi
counts.

```{r}
umi_df <- as.data.frame(cbind(sce$total, sce$cluster_k5))
colnames(umi_df) <- c("umi", "cluster_k5")

# caculate the percentatges of cells that are "low umi"
umi_df <- 
  umi_df %>% 
  mutate(low_umi = umi < min_umi) %>% 
  group_by(cluster_k5) %>% 
  summarise(n_low_umi = sum(low_umi), n_cells = n()) %>%
  mutate(pct_cell_low_umi = (n_low_umi/n_cells)*100) 
umi_df %>% arrange(desc(pct_cell_low_umi))

# Filter the clusters that have a high proportion of "low umi"
low_umi_clusters <- 
  umi_df %>% 
  filter(pct_cell_low_umi > pct_cells) %>% 
  .$cluster_k5
```

The clusters flagged are `r low_umi_clusters`

### Mithocondrial genes

High mithocondrial genes is associated with stressed, lower quality, cells.

```{r}
plotTSNE(sce, colour_by = "subsets_mt_percent") + 
  ggtitle("Percentatge mithocondrial genes")
```

```{r}
pct_cells<- 50
pct_mt  <- 10
```

Select clusters with `r pct_cells` % cells having more than `r pct_mt`%
mithocondrial genes.

```{r}
mt_df <- as.data.frame(cbind(sce$subsets_mt_percent, sce$cluster_k5))
colnames(mt_df) <- c("subsets_mt_percent", "cluster_k5")

mt_df <- 
  mt_df %>% 
  mutate(high_pct = subsets_mt_percent > pct_mt) %>% 
  group_by(cluster_k5) %>% 
  summarise(n_high_mt = sum(high_pct), n_cells = n()) %>% 
  mutate(pct_cell_high_mt = (n_high_mt/n_cells)*100)
mt_df %>% arrange(desc(pct_cell_high_mt))

high_mt_clusters <- 
  mt_df %>% 
  filter(pct_cell_high_mt > pct_cells) %>% 
  .$cluster_k5

```

The clusters flagged are `r high_mt_clusters`.

### Ribosomal genes

During the first annotation a cluster was highlighted as most of the marker
genes (from highly expressed and DE with other clusters) are ribosomal genes
(cluster 10 at k60, cluster 12 k5). As advised [in this github
comment](https://github.com/satijalab/seurat/issues/759#issuecomment-643784886)
the cluster will be removed.

To further visualise the ribosomal content in the whole dataset we plotted the
cells according to their ribosomal content.

```{r}
# the cluster 
# save ribosomal genes
is_ribo <- grepl("^Rp[sl]", rownames(sce))
# recompute the quality metrics, subseting the ribosomal genes
ribo_qc <- perCellQCMetrics(sce, subsets = list(ribo = is_ribo))
# we are only interested in ribo
sce$subsets_ribo_percent <- ribo_qc$subsets_ribo_percent
sce$subsets_ribo_sum <- ribo_qc$subsets_ribo_sum
sce$subsets_ribo_detected <- ribo_qc$subsets_ribo_detected

```

```{r}
plotTSNE(sce, colour_by = "subsets_ribo_percent") + 
  ggtitle("Percentatge mithocondrial genes")

sce$ribo_high <- ifelse(sce$subsets_ribo_percent > 10, TRUE, FALSE)
plotTSNE(sce, colour_by = "ribo_high") + 
  ggtitle("Percentatge mithocondrial genes")
```

The left most cluster is cluster 10 at k60, 12 at k5, that will removed as it
expresses markers from different celltypes and seem to only be driven by its
ribosomal content (more information in annotation_01).

```{r}
ribo_cluster <- "2"

```

### Number of mice per cluster

How many mice contribute to each cluster?

```{r mice-per-cluster}
# count how many cells from each gnt group  there are per cluster
sum_per_gnt_cluster <- table(sce$genotype, sce$cluster_k5 )
# for each cluster count how many cells are from each mice, dividing by KO and WT mice
sum_per_mice_cluster <- table(sce$mouse, sce$genotype, sce$cluster_k5 )
# For each cluster sum of mice that do have cells on that cluster
colSums(sum_per_mice_cluster > 0)

# create a summary 
summary <- as.data.frame(rbind(colSums(sum_per_mice_cluster > 0), sum_per_gnt_cluster, colSums(sum_per_gnt_cluster)))
row.names(summary) <- c("KO mice", "WT mice", "KO cells", "WT cells", "total cells")
summary
```

Except from the obvious microglia clusters, where the numbers are very low or
even absent in the fire mice nothing stands out.

### Clusters of doublet cells

Detection of clusters formed by doublets/multiplets (i.e. multiple cells
captured within the same droplet or reaction volume). The function test each
cluster against the null hypothesis that it *does* consist of doublets. The null
is rejected if a cluster has many DE genes that lie outside the expression
limits defined by the "source" clusters.

```{r}
if(!file.exists(here("outs",age,  "doublet_clusters_k5_res.csv"))){
  res_dbl <- findDoubletClusters(sce, sce$cluster_k5)
  res_dbl <- res_dbl %>% 
    as.data.frame() %>% 
    rownames_to_column("cluster")
  write.csv(res_dbl, here("outs", age,  "doublet_clusters_k5_res.csv"))
}else{
  res_dbl <- read.csv(here("outs", age,  "doublet_clusters_k5_res.csv"), row.names = 1)
}
```

Analyse the results, this includes:

-   filtering for the clusters where the null hypothesis (of being a doublet)
    was not rejected at a significance level of 5%

-   filter for the clusters that are in average formed by cells with bigger
    library sizes than its source clusters ( the umi counts of a doublet is
    expected to be larger than for a single cell).

-   Finally plot the clusters that were not already detected as poorer quality,
    along with the source clusters to have a closer look at these potential
    doublet cells.

```{r}
# filter rows where the null hypothesis was not rejected
dbl_exploratory <- res_dbl %>% 
  filter(p.value > 0.05) %>% 
# Filter the ones that were not already known as poorer quality
  filter(!(cluster %in% c( high_mt_clusters, low_umi_clusters))) %>% 
  # filter for a ratio of lib size lower than 1 ( the doublets should have a bigger lib.size)
  filter(lib.size1 < 1 & lib.size2 < 1)


#  see which ones there are and from which they come from 
dbl_exploratory
sce$dbl_exploratory <- ifelse(sce$cluster_k5 %in% dbl_exploratory$cluster, TRUE, FALSE)
plotTSNE(sce[,sce$cluster_k5 %in% c(dbl_exploratory$cluster, dbl_exploratory$source1, dbl_exploratory$source2)], text_by = "cluster_k5", colour_by = "dbl_exploratory" )

```

Cluster 20 is detected as a doublet cluster between 45 and 24. All these three
clusters are Oligodendrocytes, this "doublet" cluster might simply be to close
similarity between the clusters.

Cluster 31 is potentially only a subtype of Endothelial/Pericyte cells, as it is
flagged as a possible doublet from 29 (Endothelial/Pericyte) and 6 ( low quality
cluster, unknown cell type).

Cluster 9: is a doublet between Astrocytes (17) and Microglia(30).

53: is a doublet between Oligodendrocytes (21) and microglia (11).

```{r}
doublets <- c(9, 53)
```

### Control vs fire mice

We want to have a closer look at the clusters that do have a difference between
the knockout and the wild type before deleting the clusters.

```{r fig.width=10, message=FALSE}
# divide the two objects
sce_ko <- sce[,sce$genotype == "KO"]
sce_ctr <- sce[,sce$genotype == "WT"]
# plot them side by side
gridExtra::grid.arrange(
plotTSNE(sce_ko, colour_by = "cluster_k5", point_size=0.5,
         point_alpha = 0.3, text_by = "cluster_k5", text_size = 3) +
  scale_color_hue() +
  ggtitle("fire mice"),
plotTSNE(sce_ctr, colour_by = "cluster_k5", point_size=0.5,
         point_alpha = 0.3, text_by = "cluster_k5", text_size = 3) +
  scale_color_hue() +
  ggtitle("control"), 
ncol = 2
)
```

#### Proportion KO-WT

In order to visualise the proportions from KO and WT for each cluster, we do not
take in consideration the microglia clusters, as these are only present in the
control, and we normalise per number of cells per cluster.

```{r}
# delete the microglia
sce_no_mc <- sce[,!(sce$cluster_k5 %in% c(5, 41, 11, 30, 9, 53, 11, 9, 54))]

# count how many cells from each gnt group  there are per cluster
sum_per_gnt_cluster_no_mc <- table(sce_no_mc$genotype, sce_no_mc$cluster_k5 )

# normalise per cluster, looking how the KO and WT are distributed
# across the clusters, to give to both groups the same weight
prop_per_cluster_no_mc <- prop.table(sum_per_gnt_cluster_no_mc, margin = 1)

# calculate the proportions for each cluster
prop_per_gnt_no_mc <- round(prop.table(prop_per_cluster_no_mc , margin = 2 )*100, 2)

# Display
prop_per_gnt_no_mc<- as.data.frame(prop_per_gnt_no_mc)
colnames(prop_per_gnt_no_mc) <- c("Genotype", "cluster", "Proportion")

prop_per_gnt_no_mc %>% 
  filter(Genotype == "KO") %>% 
  arrange(desc(Proportion))

```

visualise in a plot

```{r plot}
ggplot(data = prop_per_gnt_no_mc, aes(x = cluster, y = Proportion, fill = Genotype)) +
  geom_bar(position = "fill", stat = "identity") + theme_classic() 

```

```{r}
ko_pct <- 60
```

```{r}
difference_KO_WT <-
  prop_per_gnt_no_mc %>% 
  arrange(desc(Proportion)) %>% 
# the microglia clusters will be the top differences, but this is not 
    # something we are specially interested
  filter(!(cluster %in% c(21, 24, 1, 16))) %>% 
    filter(Genotype == "KO") %>% 
    # select the "interesting" clusters
  filter(Proportion > ko_pct ) %>% 
  .$cluster
```

The clusters with more than `r ko_pct` % cells from the KO animals that were
previously flagged as lower quality are:

```{r}
gt_ko_pct <- unique(sce$cluster_k5[sce$cluster_k5 %in% c( high_mt_clusters, low_umi_clusters) &
    (sce$cluster_k5 %in% difference_KO_WT)])
```

`r gt_ko_pct`

6 and 27 are very low quality non identified cells, most likely only clustering
due to the quality and will be deleted.

37 is close to the big high mt. oligo cluster we are deleting. It does not have
high mt but does have very low umi counts. This will also be deleted.

32 and 23 are two small neuronal clusters (part of cluster 15 at k60) (approx.
70% KO). The other small neuronal clusters around, 36 and 12, would be deleted
with the chosen thresholds (approx. 55% KO). However as cluster 15 (at k60) is
the only neuronal cluster this whole cluster will be kept even though it is
borderline in terms of quality. More information supporting this choice can be
found in the cluster_QC_k60 and the first annotation files.

```{r}
# the cluster 15 from k60 is formed by the clusters bellow at k5 (cluster tree in clustering_01)
neuro <- c(12,32,23,35,36)
```

### Cluster QC

We filter out the clusters highlighted as:

-   low umi: majority of cells having less than 3000 umi counts

-   high mt : majority of cells having more than 10 % mitochondrial genes

-   doublets: their expression profile lie between two other "source" clusters,
    from which it is not expected to have intermediate cell types.

-   ribosomal content: cluster which expression is mixed between different
    celltypes and is only driven by the ribosomal content.

```{r filterout}
sce$filter_out <- sce$cluster_k5 %in% c( high_mt_clusters, low_umi_clusters, doublets, ribo_cluster) &
                !(sce$cluster_k5 %in% neuro)
filter_out <- sum(sce$filter_out)
plotTSNE(sce, colour_by = "filter_out", point_size=0.5,
         point_alpha = 0.3, text_by = "cluster_k5", text_size = 3) 

if (!file.exists(here("processed", age, "sce_clusterQC.RDS"))){
 sce <- sce[sce$filter_out == FALSE, ]
 saveRDS(sce, here("processed", age, "sce_clusterQC.RDS"))
}


```

There are `r filter_out` cells filtered out and the final object has
`r sum(sce$filter_out == FALSE)`cells.

## Session Info

<details>

<summary>

Click to expand

</summary>

```{r session-info}
sessionInfo()
```

</details>
