---
title: "Cluster quality control"
author: "Nadine Bestard"
date: "28/05/2021"
output:
  html_document:
    df_print: paged
editor_options: 
  markdown: 
    wrap: 80
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### set-up

```{r set-up, message=FALSE, warning=FALSE}
library(SingleCellExperiment)
library(here) # reproducible paths
library(scater) # plot reduced dims
library(dplyr) # df filtering
library(tibble)  # for `rownames_to_column`
library(scDblFinder) # detect doublets

```

```{r age}
age <- "old"
```

```{r load}
sce <- readRDS(here("processed", age, "sce_clusters.RDS"))
```

### Introduction

As in [publication] we will perform a cluster QC to remove clusters of poorer
quality. This will be assessed by the number of UMI counts, the mitochondrial
percentage and the number of mice that contribute to each cluster. To do so we
use a small cluster resolution, 5

```{r}
plotTSNE(sce, colour_by = "cluster_k5", point_size=0.5,
         point_alpha = 0.3, text_by = "cluster_k5", text_size = 3) +
  scale_color_hue() 
```

### Number of molecules per cluster

```{r}
plotTSNE(sce, colour_by = "total") + 
  ggtitle("Total number of umi counts per cell")
plotTSNE(sce, colour_by = "detected") +
  ggtitle("Detected number of genes per cell")
```

Lower values of umi counts and detected genes can be associated to lower quality
cells. Cells can also have lower expressed genes due to their biological state
or celltype.

```{r}
pct_cells<- 50
min_umi  <- 3000
```

Select clusters with `r pct_cells` % cells having less than `r min_umi`umi
counts.

```{r}
umi_df <- as.data.frame(cbind(sce$total, sce$cluster_k5))
colnames(umi_df) <- c("umi", "cluster_k5")

# caculate the percentatges of cells that are "low umi"
umi_df <- 
  umi_df %>% 
  mutate(low_umi = umi < min_umi) %>% 
  group_by(cluster_k5) %>% 
  summarise(n_low_umi = sum(low_umi), n_cells = n()) %>%
  mutate(pct_cell_low_umi = (n_low_umi/n_cells)*100) 
umi_df %>% arrange(desc(pct_cell_low_umi))

# Filter the clusters that have a high proportion of "low umi"
low_umi_clusters <- 
  umi_df %>% 
  filter(pct_cell_low_umi > pct_cells) %>% 
  .$cluster_k5
```

The clusters flagged are `r low_umi_clusters`

### Mithocondrial genes

High mithocondrial genes is associated with stressed, lower quality, cells.

```{r}
plotTSNE(sce, colour_by = "subsets_mt_percent") + 
  ggtitle("Percentatge mithocondrial genes")
```

```{r}
pct_cells<- 50
pct_mt  <- 10
```

Select clusters with `r pct_cells` % cells having more than `r pct_mt`%
mithocondrial genes.

```{r}
mt_df <- as.data.frame(cbind(sce$subsets_mt_percent, sce$cluster_k5))
colnames(mt_df) <- c("subsets_mt_percent", "cluster_k5")

mt_df <- 
  mt_df %>% 
  mutate(high_pct = subsets_mt_percent > pct_mt) %>% 
  group_by(cluster_k5) %>% 
  summarise(n_high_mt = sum(high_pct), n_cells = n()) %>% 
  mutate(pct_cell_high_mt = (n_high_mt/n_cells)*100)
mt_df %>% arrange(desc(pct_cell_high_mt))

high_mt_clusters <- 
  mt_df %>% 
  filter(pct_cell_high_mt > pct_cells) %>% 
  .$cluster_k5

```

The clusters flagged are `r high_mt_clusters`.


### Number of mice per cluster

How many mice contribute to each cluster?

```{r mice-per-cluster}
# count how many cells from each gnt group  there are per cluster
sum_per_gnt_cluster <- table(sce$genotype, sce$cluster_k5 )
# for each cluster count how many cells are from each mice, dividing by KO and WT mice
sum_per_mice_cluster <- table(sce$mouse, sce$genotype, sce$cluster_k5 )
# For each cluster sum of mice that do have cells on that cluster
colSums(sum_per_mice_cluster > 0)

# create a summary 
summary <- as.data.frame(rbind(colSums(sum_per_mice_cluster > 0), sum_per_gnt_cluster, colSums(sum_per_gnt_cluster)))
row.names(summary) <- c("KO mice", "WT mice", "KO cells", "WT cells", "total cells")
summary
```

Except from the obvious microglia clusters, where the numbers are very low or
even absent in the fire mice nothing stands out.

### Clusters of doublet cells

Detection of clusters formed by doublets/multiplets (i.e. multiple cells captured within the same droplet or reaction volume). The function test each cluster against the null hypothesis that it _does_ consist of doublets. The null is rejected if a cluster has many DE genes that lie outside the expression limits defined by the “source” clusters.

```{r}
if(!file.exists(here("outs",age,  "doublet_clusters_k5_res.csv"))){
  res_dbl <- findDoubletClusters(sce, sce$cluster_k5)
  res_dbl <- res_dbl %>% 
    as.data.frame() %>% 
    rownames_to_column("cluster")
  write.csv(res_dbl, here("outs", age,  "doublet_clusters_k5_res.csv"))
}else{
  res_dbl <- read.csv(here("outs", age,  "doublet_clusters_k5_res.csv"), row.names = 1)
}
```

Analyse the results, this includes filtering for the clusters where the null hypothesis (of being a doublet) was not rejected at a significance level of 5%, filter for the ones that are in average formed by cells with bigger library sizes than its source clusters ( the umi counts of a doublet is expected to be larger than for a single cell). Finally plot the clusters that were not already detected as poorer quality, along with the source clusters to have a closer look at these potential doublet cells. 

```{r}
# filter rows where the null hypothesis was not rejected
dbl_exploratory <- res_dbl %>% 
  filter(p.value > 0.05) %>% 
# Filter the ones that were not already known as poorer quality
  filter(!(cluster %in% c( high_mt_clusters, low_umi_clusters))) %>% 
  # filter for a ratio of lib size lower than 1 ( the doublets should have a bigger lib.size)
  filter(lib.size1 < 1 & lib.size2 < 1)


#  see which ones there are and from which they come from 
dbl_exploratory
sce$dbl_exploratory <- ifelse(sce$cluster_k5 %in% dbl_exploratory$cluster, TRUE, FALSE)
plotTSNE(sce[,sce$cluster_k5 %in% c(dbl_exploratory$cluster, dbl_exploratory$source1, dbl_exploratory$source2)], text_by = "cluster_k5", colour_by = "dbl_exploratory" )

```



Cluster 20 is detected as a doublet cluster between 45 and 24. The cluster 20 is clustered with 45 in the next resolution, therfore it might just be due to the close similarity between this two clusters. 
The other 3 clusters are doublets of:
9: Astrocytes and Microglia, 
31: Endothelial cells and unknown
53: Oligodendrocytes and microglia. 

**TODO:** check with David 9 and 53 are doublets, and there is not known celltypes that could express combinations of these celltypes.

```{r}
doublets <- c(9, 31, 53)
```


### Control vs fire mice

We want to have a closer look at the clusters that do have a difference between the knockout and the wild type before deleting the clusters.

```{r fig.width=10, message=FALSE}
# divide the two objects
sce_ko <- sce[,sce$genotype == "KO"]
sce_ctr <- sce[,sce$genotype == "WT"]
# plot them side by side
gridExtra::grid.arrange(
plotTSNE(sce_ko, colour_by = "cluster_k5", point_size=0.5,
         point_alpha = 0.3, text_by = "cluster_k5", text_size = 3) +
  scale_color_hue() +
  ggtitle("fire mice"),
plotTSNE(sce_ctr, colour_by = "cluster_k5", point_size=0.5,
         point_alpha = 0.3, text_by = "cluster_k5", text_size = 3) +
  scale_color_hue() +
  ggtitle("control"), 
ncol = 2
)
```

#### Proportion KO-WT

In order to visualise the proportions from KO and WT for each cluster, we
do not take in consideration the microglia clusters, as these are only present in the control, and we normalise per number
of cells per cluster.

```{r}
# delete the microglia
sce_no_mc <- sce[,!(sce$cluster_k5 %in% c(5, 41, 11, 30, 9, 53, 11, 9, 54))]

# count how many cells from each gnt group  there are per cluster
sum_per_gnt_cluster_no_mc <- table(sce_no_mc$genotype, sce_no_mc$cluster_k5 )

# normalise per cluster, looking how the KO and WT are distributed
# across the clusters, to give to both groups the same weight
prop_per_cluster_no_mc <- prop.table(sum_per_gnt_cluster_no_mc, margin = 1)

# calculate the proportions for each cluster
prop_per_gnt_no_mc <- round(prop.table(prop_per_cluster_no_mc , margin = 2 )*100, 2)

# Display
prop_per_gnt_no_mc<- as.data.frame(prop_per_gnt_no_mc)
colnames(prop_per_gnt_no_mc) <- c("Genotype", "cluster", "Proportion")

prop_per_gnt_no_mc %>% 
  filter(Genotype == "KO") %>% 
  arrange(desc(Proportion))

```

visualise in a plot


```{r plot}
ggplot(data = prop_per_gnt_no_mc, aes(x = cluster, y = Proportion, fill = Genotype)) +
  geom_bar(position = "fill", stat = "identity") + theme_classic() 

```

```{r}
ko_pct <- 60
```

```{r}
difference_KO_WT <-
  prop_per_gnt_no_mc %>% 
  arrange(desc(Proportion)) %>% 
# the microglia clusters will be the top differences, but this is not 
    # something we are specially interested
  filter(!(cluster %in% c(21, 24, 1, 16))) %>% 
    filter(Genotype == "KO") %>% 
    # select the "interesting" clusters
  filter(Proportion > ko_pct ) %>% 
  .$cluster
```

The clusters with more than `r ko_pct` % cells from the KO animals that were previously flagged as lower quality are:

```{r}
gt_ko_pct <- unique(sce$cluster_k5[sce$cluster_k5 %in% c( high_mt_clusters, low_umi_clusters) &
    (sce$cluster_k5 %in% difference_KO_WT)])
```

`r gt_ko_pct`


 6 and 27 are very low quality non identified cells, most likely only clustering due to the quality and will be deleted. 
 
 37 is close to the big high mt. oligo cluster we are deleting. It does not have high mt but does have very low umi counts. This will also be deleted. 
 
32 and 23 are two small neuronal clusters (part of cluster 15 at k60) (approx. 70% KO). The other small neuronal clusters around, 36 and 12, would be deleted with the chosen thresholds (approx. 55% KO). However as cluster 15 (at k60) is the only neuronal cluster this whole cluster will be kept even though it is borderline in terms of quality. More information supporting this choice can be found in the cluster_QC_k60 and the first annotation files.

```{r}
# the cluster 15 from k60 is formed by the clusters bellow at k5 (cluster tree in clustering_01)
neuro <- c(12,32,23,35,36)
```

### Cluster QC

We filter out the clusters highlighted as:

-   low umi: majority of cells having less than 3000 umi counts

-   high mt : majority of cells having more than 10 % mitochondrial genes

-   doublets: their expression profile lie between two other “source” clusters.
**TODO:** check with David 9 and 53 are doublets, and there is not known celltypes that could express these combinations.

```{r filterout}
sce$filter_out <- sce$cluster_k5 %in% c( high_mt_clusters, low_umi_clusters, doublets) &
                !(sce$cluster_k5 %in% neuro)
filter_out <- sum(sce$filter_out)
plotTSNE(sce, colour_by = "filter_out", point_size=0.5,
         point_alpha = 0.3, text_by = "cluster_k5", text_size = 3) 

if (!file.exists(here("processed", age, "sce_clusterQC.RDS"))){
 sce <- sce[sce$filter_out == FALSE, ]
 saveRDS(sce, here("processed", age, "sce_clusterQC.RDS"))
}


```

There are `r filter_out` cells filtered out.

## Session Info

<details>

<summary>

Click to expand

</summary>

```{r session-info}
sessionInfo()
```

</details>
