---
title: "Differential Expression"
author: "Nadine Bestard"
date: "21/07/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

#### set-up
```{r}
library(scran) # for scDE
library(scater) # for aggregate counts
library(edgeR) #for De
library(here) # reproducible paths
```
```{r}
age <- "old"
sce <- readRDS(here("processed", age, "sce_anno_02.RDS"))
```


## Use of pseudo-bulk samples

We perform the DE analysis separately for each label to identify cell type-specific transcriptional effects of KO condition. The actual DE testing is performed on “pseudo-bulk” expression profiles (Tung et al. 2017), generated by summing counts together for all cells with the same combination of label and sample. This leverages the resolution offered by single-cell technologies to define the labels, and combines it with the statistical rigor of existing methods for DE analyses involving a small number of samples.

```{r}
# each column represents unique combination for clusters_named and each mouse replicate
summed <- aggregateAcrossCells(sce, 
    id=colData(sce)[,c("clusters_named", "mouse")])
summed
```

## Perform DE

We first filter all the combinations sample-label that resulted in less than 10
cells. These are saved in /outs/`r age`/DE/fail_min_cells_DE.csv

```{r}
# Removing all pseudo-bulk samples with 'insufficient' cells.
summed_filt <- summed[, summed$ncells >= 10]

# and saving to file which one there are
fail_n_cells  <- colData(summed)[ colData(summed)$ncells < 10, c("Sample", "mouse", "genotype", "chip", "clusters_named", "ncells")]
write.csv(fail_n_cells, here("outs", age,"DE", "fail_min_cells_DE.csv"), row.names = FALSE)
```

We then compute the DE with edgeR, one of the best methods according to [(Soneson et al. 2018)](https://www.nature.com/articles/nmeth.4612). The results are saved in /outs/`r age`/DE/de_results

If the comparison was not possible for a particular cluster, most commonly due to lack of residual degrees of freedom ( e.g. an absence of enough samples from both conditions) it is listed in /outs/`r age`/DE/fail_min_cells_DE.csv

```{r}
# Reordering the genotype factor, treated should be second level
summed_filt$genotype <- factor(summed_filt$genotype, levels = c("WT", "KO"))

# run de
de_results <- pseudoBulkDGE(summed_filt, 
    label=summed_filt$clusters_named,
    design= ~factor(chip) + genotype,
    coef="genotypeKO",
    condition=summed_filt$genotype 
)

# save result
lapply(names(de_results), function(x){
  de_results_clu <- de_results[[x]]
  write.csv(de_results_clu, here("outs", age, "DE", "de_results", paste0("de_results_", x, ".csv")), quote = FALSE)
}
)

# save fail
write.csv(metadata(de_results), here("outs", age,"DE", "fail_contrast_DE.csv"), row.names = FALSE)

```

